parsed.go

```diff
@@ -0,0 +1,353 @@
+package proteus
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/simplesurance/proteus/types"
+)
+
+const redactedPlaceholder = "<redacted>"
+
+// Parsed holds information about all parameters supported by the application,
+// and their options, allowing interacting with them.
+type Parsed struct {
+	settings      settings
+	inferedConfig config
+	valuesMutex   sync.Mutex
+	values        []types.ParamValues
+}
+
+// Usage print usage information to the provider writer.
+func (p *Parsed) Usage(w io.Writer) {
+	p.usage(w, nil)
+}
+
+// ErrUsage is a specialized version of Usage(), that is intended
+// to be used when the wrong parameters where provided to the application, and
+// will show to the user what is wrong with the provided parameters.
+// It does not terminate the application.
+func (p *Parsed) ErrUsage(w io.Writer, err error) {
+	fmt.Fprintln(w, "Invalid configuration parameters for application")
+	if err != nil {
+		fmt.Fprintln(w, err.Error())
+	}
+
+	p.usage(w, err)
+}
+
+func (p *Parsed) usage(w io.Writer, err error) {
+	setKeys := make([]string, 0, len(p.inferedConfig))
+	for k := range p.inferedConfig {
+		setKeys = append(setKeys, k)
+	}
+
+	sort.Strings(setKeys)
+
+	details := strings.Builder{}
+	cmdLine := []string{binaryName()}
+
+	lastSet := ""
+	for _, setName := range setKeys {
+		set := p.inferedConfig[setName]
+
+		if lastSet != setName {
+			lastSet = setName
+			cmdLine = append(cmdLine, setName)
+		}
+
+		if len(set.fields) == 0 {
+			continue
+		}
+
+		fmt.Fprintln(&details)
+		if setName == "" {
+			fmt.Fprintln(&details, "PARAMETERS")
+		} else {
+			fmt.Fprintln(&details, "PARAMETER SET: "+strings.ToUpper(setName))
+			if set.desc != "" {
+				fmt.Fprintln(&details, set.desc)
+			}
+		}
+
+		// sort by parameter names
+		paramNames := make([]string, 0, len(set.fields))
+		for k := range set.fields {
+			paramNames = append(paramNames, k)
+		}
+
+		sort.Slice(paramNames, func(i, j int) bool {
+			p1 := set.fields[paramNames[i]]
+			p2 := set.fields[paramNames[j]]
+
+			// mandatory fields first
+			if p1.optional != p2.optional {
+				return p2.optional
+			}
+
+			// then lexicographic order
+			return paramNames[i] < paramNames[j]
+		})
+
+		// describe parameter name, type and options
+		for _, name := range paramNames {
+			field := set.fields[name]
+
+			cmdLine = append(cmdLine, "  "+formatCmdLineParam(name, field))
+
+			opts := []string{fmt.Sprintf("- %s:%s", name, field.typ)}
```

Comment:
To my eyes it looks silly documenting port number as `-port <PORT>`. Imo `-port <uint16>` is way better, and the advantage is even bigger for xtypes, like `-server <url.URL> -region <EU|JP>`. To avoid having the duplication of the type you mentioned on your comment, I removed the second appearance. Please have a look.